Ylz potential

Tobias-Dwyer opened the pull request:
<!-- Please confirm that your work is based on the correct branch. -->
<!-- Bug fixes should be based on *trunk-patch*. -->
<!-- Backwards compatible new features should be based on *trunk-minor*. -->
<!-- Incompatible API changes should be based on *trunk-major*. -->

## Description

<!-- Describe your changes in detail. -->
I added an isotropic pair potential known as the YLZ potential, from this publication. 10.1103/PhysRevE.82.011905
I also added documentation that follows the patchy potential format and is tested via updated Sybil format.

## Motivation and context
To simulate fluidized membranes with a pair potential allows for simulation of phenomena not captured with faster mesh potentials in a computationally efficient manner. The YLZ potential is also implemented in LAMMPS but adding this potential to HOOMD will allow additional flexibility with hoomd.  


<!--- Why is this change required? What problem does it solve? -->
This adds a commonly used potential for modelling biological systems to HOOMD. 

<!-- Replace ??? with the issue number that this pull request resolves. -->

## How has this been tested?

<!--- Please describe in detail how you tested your changes. -->
I followed the example  testing infrastructure built for the patchy potential. I independently calculated several configurations of the potential with my own python script within each r_cut of the potential with a variety of relative orientations of the particles. In each configuration I tested the forces, torques, and energies against my independently calculated values. I tested what happens if the particles have the same symmetry vector and if they have different vectors. I think this covers all of the potential force scenarios the particles should experience. 

<!--- Please build the sphinx documentation and check that any changes to
      documentation display properly. -->
I successfully built the sphinx documentation and changes display properly. The pytests which incorporate sybil also successfully pass.

## Checklist:

- [x] I have reviewed the [**Contributor Guidelines**](https://github.com/glotzerlab/hoomd-blue/blob/trunk-minor/CONTRIBUTING.rst).
- [x] I agree with the terms of the [**HOOMD-blue Contributor Agreement**](https://github.com/glotzerlab/hoomd-blue/blob/trunk-minor/ContributorAgreement.md).
- [x] My name is on the list of contributors (`sphinx-doc/credits.rst`) in the pull request source branch.
- [x] I have summarized these changes in `CHANGELOG.rst` following the established format.

Commit:
ylz potential

Commit:
formatting

Tobias-Dwyer committed:
documentation

Tobias-Dwyer committed:
inherited

Tobias-Dwyer committed:
documentation

Tobias-Dwyer committed:
clag format

MartinGirard commented:
Couple of thoughts for consistency within hoomd. This implements the anisotropic potential, but we lack the isotropic version of it. 

The paper you listed has additional parameter space, specifically on the cosine power. Should we allow arbitrary (integer) powers here, instead of setting 2xi it to 8?

The functional form of the isotropic part is similar to: https://doi.org/10.1063/1.5089417 (although it is a 12-6 on the repulsive side). 

(Why is it YLZ and not YHLLZ?)

Tobias-Dwyer commented:
Hi @MartinGirard, thank you for your comment! 

> Couple of thoughts for consistency within hoomd. This implements the anisotropic potential, but we lack the isotropic version of it.

I agree, however for now I'm going to leave the isotropic part for future developers, I agree a cosine potential would be nice to have as a standalone for sure, as well as the full isotropic form of this potential. I'll raise this as an issue after merging this PR and see if I get to it in the coming months. 

> 
> The paper you listed has additional parameter space, specifically on the cosine power. Should we allow arbitrary (integer) powers here, instead of setting 2xi it to 8?

Yes it does, I was a bit lazy and only implemented a single cosine power - however @joaander pointed me to your work calculating exponents for hoomd math, and I'll add the functionality for the cosine power. 

> 
> The functional form of the isotropic part is similar to: https://doi.org/10.1063/1.5089417 (although it is a 12-6 on the repulsive side).
> 
> (Why is it YLZ and not YHLLZ?)

I've gotten this question several times, the implementation provided in lammps by the first author of the original paper for the potential calls it the (Yuan-Li-Zhang) potential. 

https://doi.org/10.1016/j.cpc.2016.09.018
10.1103/PhysRevE.82.011905
https://docs.lammps.org/pair_ylz.html

I ofc don't want to offend anyone, but I think keeping consistency between the potentials would be good.

Tobias-Dwyer committed:
twozeta

Tobias-Dwyer committed:
added tests

Tobias-Dwyer committed:
documentation

Tobias-Dwyer committed:
syntax

Tobias-Dwyer commented:
I added functionality to make exponent zeta a settable parameter in the pair potential, tested the range of zeta parameters from the original paper selecting twozeta values between 4 and 20. 

I updated and built the documentation.

Tobias-Dwyer commented:
The YLZ potential works to simulate a vesicle as intended. Simulated at the fluidized parameter range of the original paper reproducing qualitatively the observed results. 2zeta = 8, and kT=0.2, beta=3 and phi=0 (per the hoomd defined parameters)

https://github.com/user-attachments/assets/d96d9d31-b441-4952-a2ba-a0e68937d900

Note positions were initialized on a Fibonacci sphere.

joaander commented on hoomd/md/pair/aniso.py:
```diff
@@ -144,6 +144,118 @@ def __init__(self, nlist, default_r_cut=None):
+
+        .. code-block:: python
+
+            ylz.mu['A'] = (1.0,0.0,0.0)
```
```suggestion
            ylz.mu['A'] = (1.0, 0.0, 0.0)
```

joaander commented on hoomd/md/pair/aniso.py:
```diff
@@ -144,6 +144,118 @@ def __init__(self, nlist, default_r_cut=None):
+                        'rmin':1.122, 'twozeta': int(4)}
+
+        ylz.params.default = ylz_params
+        ylz.mu.default = (0,0,1)
```
```suggestion
        ylz.mu.default = (0, 0, 1)
```

joaander commented on hoomd/md/pair/aniso.py:
```diff
@@ -144,6 +144,118 @@ def __init__(self, nlist, default_r_cut=None):
+        ylz.mu.default = (0,0,1)
+        simulation.operations.integrator.forces = [ylz]
+
+    {inherited}
```
```suggestion
    {inherited}
    
    ----------

    **Members defined in** `YLZ`:    
```

joaander commented on hoomd/md/pair/aniso.py:
```diff
@@ -144,6 +144,118 @@ def __init__(self, nlist, default_r_cut=None):
+    orientation of a pair of particles as function of their axis of
+    symmetry :math:`\mu` in their local reference frame.
+
+    .. invisible-code-block: python
```
```suggestion
    .. code-block: python
```
We should show this example code to the users.

joaander reviewed (changes requested):
Thanks! I have some suggestions to improve the documentation.

Please also add an entry to the change log.

Tobias-Dwyer committed:
Update hoomd/md/pair/aniso.py

Co-authored-by: Joshua A. Anderson <joaander@umich.edu>

Tobias-Dwyer committed:
Update hoomd/md/pair/aniso.py

Co-authored-by: Joshua A. Anderson <joaander@umich.edu>

Tobias-Dwyer committed:
Update hoomd/md/pair/aniso.py

Co-authored-by: Joshua A. Anderson <joaander@umich.edu>

Tobias-Dwyer committed:
Update hoomd/md/pair/aniso.py

Co-authored-by: Joshua A. Anderson <joaander@umich.edu>

Tobias-Dwyer committed:
pre-commit

Tobias-Dwyer committed:
updated changelog

Tobias-Dwyer committed:
pre-commit

Tobias-Dwyer commented:
@joaander I added your documentation change requests and updated the changelog. To do this, I copied the current change log and added my contribution (not sure if this is the right way).

joaander commented:
> @joaander I added your documentation change requests and updated the changelog. To do this, I copied the current change log and added my contribution (not sure if this is the right way).

No, this is not the right way. You should merge `trunk-minor` up into your branch. Then add text specific only to the new feature.

The may detect your copy and merge cleanly, or it may generate merge conflicts. If you can't resolve the conflicts, you can revert your changes and then try merging again.

Tobias-Dwyer committed:
revert changelog

Tobias-Dwyer committed:
Merge branch 'trunk-minor' into YLZ_potential

Tobias-Dwyer committed:
merged trunk-minor

joaander committed:
Merge branch 'trunk-minor' into YLZ_potential

joaander committed:
Fix submodules.

joaander committed:
Align zetterling parameters.

joaander commented on hoomd/md/pair/aniso.py:
```diff
@@ -146,6 +146,122 @@ def __init__(self, nlist, default_r_cut=None):
+    orientation of a pair of particles as function of their axis of
+    symmetry :math:`\mu` in their local reference frame.
+
+    .. code-block: python
```
This code block still doesn't render. Maybe missing a newline?
```suggestion
    .. code-block: python

```

joaander commented on hoomd/md/pair/aniso.py:
```diff
@@ -146,6 +146,122 @@ def __init__(self, nlist, default_r_cut=None):
+    The second part of the isotropic potential is a Cosine potential that
+    is a function of both :math:`r_{min}` and :math:`r_{cut}`.:
+
+    .. math::
```
This math is not readable: https://hoomd-blue--2110.org.readthedocs.build/en/2110/hoomd/md/pair/aniso/ylz.html
I would suggest using multiple math blocks.

joaander commented on hoomd/md/pair/aniso.py:
```diff
@@ -146,6 +146,122 @@ def __init__(self, nlist, default_r_cut=None):
+    isotropic pair potential is composed of two parts, a 2,4 Lennard-Jones
+    potential set to cut off at the potential minimum, :math:`r_{min}`.
+    The second part of the isotropic potential is a Cosine potential that
+    is a function of both :math:`r_{min}` and :math:`r_{cut}`.:
```
Please cite the original reference (via a link to the DOI). See other references links in the docs for examples.

Tobias-Dwyer replied:
updated!

joaander commented on hoomd/md/pair/aniso.py:
```diff
@@ -146,6 +146,122 @@ def __init__(self, nlist, default_r_cut=None):
+        Type: `TypeParameter` [``particle_type``, `tuple` [`float`, `float`,
+        `float` ]]
+
+        .. rubric:: Examples:
```
```suggestion
        .. rubric:: Example:
```

joaander commented on hoomd/md/pair/aniso.py:
```diff
@@ -146,6 +146,122 @@ def __init__(self, nlist, default_r_cut=None):
+
+    The modulation function :math:`\psi` creates torques to align the
+    orientation of a pair of particles as function of their axis of
+    symmetry :math:`\mu` in their local reference frame.
```
```suggestion
    symmetry :math:`\mu` in their local reference frame.
    
    .. rubric:: Example:
```

joaander reviewed (changes requested):
There are still a number of issues with the documentation that need to be fixed before we can merge this.

Tobias-Dwyer committed:
Update hoomd/md/pair/aniso.py

Co-authored-by: Joshua A. Anderson <joaander@umich.edu>

Tobias-Dwyer committed:
Update hoomd/md/pair/aniso.py

Co-authored-by: Joshua A. Anderson <joaander@umich.edu>

Tobias-Dwyer committed:
Update hoomd/md/pair/aniso.py

Co-authored-by: Joshua A. Anderson <joaander@umich.edu>

Tobias-Dwyer committed:
documentatation

Tobias-Dwyer committed:
added nlist

Tobias-Dwyer committed:
changed nlist

joaander committed:
Fix doc indentation.

Tobias-Dwyer committed:
units

Tobias-Dwyer committed:
precommit

Tobias-Dwyer commented:
@joaander added units to the documentation, once tests pass, this should be ready to merge.

kmjens commented:
> @joaander added units to the documentation, once tests pass, this should be ready to merge.

Hi both! Currently in process of reviewing. Will finish by tomorrow evening. Thanks

kmjens committed:
Edit typos and wording of aniso.py documentation

kmjens committed:
appease pre-commit gods

kmjens commented:
When I build the docs locally for this branch, I get a "failing remark" next to the github symbol at the top. Not entirely sure what that means or if this is cause for concern. The docs otherwise look like they build properly and the pytest passes, but I just thought I'd flag it just in case.


<img width="759" height="265" alt="Screenshot 2025-08-22 at 11 59 24 PM" src="https://github.com/user-attachments/assets/26bf5d40-e21e-4194-ae51-ef7933b7505f" />

kmjens committed:
appease pre-commit gods

kmjens committed:
correction to my correction

kmjens committed:
more concise language in documentation

kmjens reviewed (approved):
Made changes to documentation to correct typos and make language more concise. Rest looks good! Excited to use this potential soon.

Tobias-Dwyer commented:
> Made changes to documentation to correct typos and make language more concise. Rest looks good! Excited to use this potential soon.

Thanks @kmjens!

joaander committed:
Remove testing badge

It gives many false negatives.

joaander commented:
> When I build the docs locally for this branch, I get a "failing remark" next to the github symbol at the top. Not entirely sure what that means or if this is cause for concern. The docs otherwise look like they build properly and the pytest passes, but I just thought I'd flag it just in case.

It means that the most recent test run on GitHub actions has failed. This is most often caused by another PR or branch with failing tests. To see the test results for a specific PR, scroll to the bottom of of the PR and expand the "Checks" section.

I'll remove the badge as it often gives false negative results.

joaander reviewed (approved):
Thanks!
